### PR TITLE
Redesign for constraints page

### PR DESCRIPTION
**Pull request by cdccollins**
### Description of change

- Add/Remove links rather than checkboxes
- Can't remove constraints identified by system
- Can search for constraints

### Story Link

https://trello.com/c/qkrP4O3J/2796-the-source-of-the-constraints

### Screenshots

screencapture-lambeth-bops-localhost-3000-planning-applications-2514-validation-constraints-2024-04-22-14_30_03

### CODE REVIEW

File: app/assets/stylesheets/_main.scss [hunk 1]
```diff
@@ -457,3 +457,13 @@ body {
 .govuk-\!-text-wrap-balance {
   text-wrap: balance;
 }
+
+.dashed-box {
+  padding: 30px;
+  border: 1px dashed #B1B4B6;
```
Comment (benjamineskola) [inline app/assets/stylesheets/_main.scss:463]: It might be better to use the govuk-colour function rather than the hex code:

```suggestion
  border: 1px dashed govuk-colour("mid-grey");
```

**Review by benjamineskola — APPROVED** [review]

File: app/controllers/planning_applications/validation/constraints_controller.rb [hunk 2]
```diff
@@ -4,46 +4,49 @@ module PlanningApplications
...
+          redirect_to planning_application_validation_constraints_path(@planning_application),
+            notice: t(".failure")
+        end
+      end
+
+      def update
+        ActiveRecord::Base.transaction do
+          @planning_application.update!(updated_address_or_boundary_geojson: true)
           @planning_application.constraints_checked!
         end
```
Comment (pixeltrix) [inline app/controllers/planning_applications/validation/constraints_controller.rb:42]: This should probably be extracted to a method on the model

File: app/controllers/planning_applications/validation/constraints_controller.rb [hunk 3]
```diff
@@ -64,22 +67,12 @@ def set_planning_application_constraints
...
-      end
-
-      def existing_constraint_ids
-        @existing_constraint_ids ||= @planning_application_constraints.map(&:constraint_id)
-      end
-
-      def constraints_to_add
-        @constraints_to_add ||= constraint_ids - existing_constraint_ids
+      def set_other_constraints
+        @other_constraints = Constraint.all_constraints(search_param).non_applicable_constraints(@planning_application.planning_application_constraints).sort_by(&:category)
```
Comment (pixeltrix) [inline app/controllers/planning_applications/validation/constraints_controller.rb:71]: This should be a method on the `Constraint` model, e.g.

```
@other_constraints = Constraint.other_constraints(search_param, @planning_application)
```

File: config/routes.rb [hunk 4]
```diff
@@ -168,7 +168,9 @@
 
           resource :reporting_type, only: %i[edit update]
 
-          resource :constraints, only: %i[show update]
+          resources :constraints, only: %i[index create destroy] do
+            patch :update, on: :collection
+          end
```
Comment (pixeltrix) [inline config/routes.rb:173]: Why the change to `resources`?

**Review by pixeltrix — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 4 above)
Comment (cdccollins) [reply config/routes.rb:173]: Because I wanted the :id on the create and destroy method?

**Review by pixeltrix — APPROVED** [review]

**Review by EGiataganas — APPROVED** [review]
